lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,423 @@
+""" Collection and Structure Module for pbench-audit-server
+"""
+
+import os
+import time
+
+from collections import OrderedDict
+
+
+permdict = {
+    "0": "---",
+    "1": "--x",
+    "2": "-w-",
+    "3": "-wx",
+    "4": "r--",
+    "5": "r-x",
+    "6": "rw-",
+    "7": "rwx",
+}
+
+
+def filepermissions(mode):
+    """Convert File permissions from numeric to symbolic.
+    from '755' --> 'rwxr-xr-x' """
+    fperm = ""
+    modebits = oct(mode)[-3:]
+    for bits in modebits:
+        if bits in permdict:
+            fperm += permdict[bits]
+        else:
+            raise ValueError(
+                f"Could not find key '{bits}' in file permissions dictionary"
+            )
+
+    return fperm
+
+
+class Hierarchy(object):
+    """Parent class of Hierarchies"""
```

Comment:
Typically one calls this the `base` class or `super` class of all hierarchy derived classe.